artifacts/definitions/Windows/EventLogs/Evtx.yaml

```diff
@@ -0,0 +1,92 @@
+  export (_e.g._, to elasticsearch) for additional processing.  It can also be 
+  used as the basis for custom artifacts with more in-depth filtering.
+  
+  **Note: This artifact can be resource intensive.**  
+  
+  - Parsing and aggregating may use high amounts of CPU on the client. Consider
+  reducing the ops/second or narrowing the glob/path regex if necessary.
+  - Parsing many event logs may take longer than the default timeout.  When
+  parsing all log files and searching VSS, consider doubling the default or 
+  more (especially with reduced ops/second, or if targets have high-volume 
+  3rd-party log sources such as Sysmon).
+  - The artifact routinely produces hundreds of thousands of rows per host. 
+  Consider filtering results using path, channel, and ID regexes if necessary.
+
+  Inspired by others in `Windows.EventLogs.*`, many by Matt Green (@mgreen27).
+    
+author: Chris Hendricks (chris@counteractive.net)
+
+precondition: SELECT OS FROM info() WHERE OS = 'windows'
+
+parameters:
+  - name: EvtxGlob
+    default: '%SystemRoot%\System32\winevt\Logs\*.evtx'
+  - name: SearchVSS
+    description: "Search VSS for EvtxGlob as well."
+    type: bool
+  - name: StartDate
+    type: timestamp
+    description: "Parse events on or after this date (YYYY-MM-DDTmm:hh:ssZ)"
+  - name: EndDate
+    type: timestamp
+    description: "Parse events on or before this date (YYYY-MM-DDTmm:hh:ssZ)"
+  - name: PathRegex
+    default: "."
+  - name: ChannelRegex
+    default: "."
+  - name: IDRegex
+    default: "."
+
+sources:
+  - queries:
+      - LET evtxglobs <= SELECT FullPath FROM glob(globs=expand(path=EvtxGlob))
+
+      - LET files = 
+          SELECT * FROM foreach(
+            row=evtxglobs,
+            query={
+              SELECT * 
+              FROM if(condition=SearchVSS,
+                      then={SELECT * FROM Artifact.Windows.Search.VSS(SearchFilesGlob=FullPath)},
+                      else={SELECT * FROM glob(globs=FullPath)})
+            }
+          )
+          WHERE FullPath =~ PathRegex
+          ORDER BY FullPath DESC
+
+      # Parse all target files and de-duplicate with GROUP BY
+      - SELECT * FROM foreach(
+          row=files,
+          query={
+            SELECT *,
+              timestamp(epoch=int(int=System.TimeCreated.SystemTime)) AS TimeCreated,
+              System.Channel as Channel,
+              System.EventRecordID as EventRecordID,
+              System.EventID.Value as EventID,
+              FullPath
+            FROM parse_evtx(filename=FullPath)
+            WHERE
+              if(condition=StartDate, then=TimeCreated >= timestamp(string=StartDate), else=true)
+              AND if(condition=EndDate, then=TimeCreated <= timestamp(string=EndDate), else=true)
+              AND Channel =~ ChannelRegex
+              AND str(str=EventID) =~ IDRegex
+          }
+        )
+        GROUP BY EventRecordID,Channel
```

Comment:
Consider the need for this group by - when you add a group by it forces VQL to queue all rows in memory (so it can deduplicate them) and this means all that data is kept in memory before forwarding to the server. As is, this artifact will keep every message in memory before sending (because the group by is on the outer query).

It is really only needed for VSS deduplication where you want to remove the same event record ids appearing in multiple VSS copies of the same file. For that use case it is possibly better to deduplicate on the level of a single log file (because then we only need to queue one log file's worth of messges at the time). 

@mgreen27 I think this might be an issue that affects all the event log artifacts? I feel that we are better off to write a generic artifact that enumerates all the events, taking care of the VSS deduplicating and then just call it from everywhere else. That artifact can just be a drop in replacement for parse_evtx() everywhere.

Perhaps dont worry about the VSS for this one?

Reply:
Certainly fair points, sir, though it works well enough in practice (and is consistent with the rest of the event log artifacts, as we based it on them :)) The VSS does give us a few percentage points of extra coverage, time wise, so it's useful in forensic contexts.

One side effect of having it on the outer query is you can target event log archives (copies of evtx files) which may contain multiple copies of overlapping files, with the same channels and event record ids.  Rare, for sure, but VSS isn't the only de-dupe situation I don't think.

Reply:
I think the impact is proportional to the total number of events - if we just want to collect all the events then storing them all in memory will be problematic. If we have a tighter filter and/or date range it might not be such a problem.

Perhaps split the artifact into two? one with a deduplicating option and one without?  So you can put the group by on only one branch but in the common case we dont have that additional cost?

Reply:
Something like this:
```
LET X = SELECT ....
SELECT * FROM if(condition= Deduplicate="Y",
   then={
      SELECT * FROM X
      GROUP BY ...
   }, else=X)
```

Reply:
For a little context (small n, of course, and tons of complicating factors), running VSS on all logs gets maybe ~3-4K unique additional records on ~300-400K events, about 1%.  For that you "pay" about 700 seconds in execution time: ~150 without VSS, ~850 with.  Y'all probably have _much_ more objective and representive stats 😃

Reply:
I am more concerned about the memory usage then the execution time. Do you have an idea of that?  You can see the memory usage in the host screen under vql drill down.

Reply:
yeah, it spikes for sure 🔥 - shows around 4-5000 on a stock install, presumably that's megabytes?  so 4GB of ram parsing maybe 400MB of on-disk logs, plus 2-3 shadow copies.

Reply:
I'll take another look tomorrow and try to work in the feedback, though your velocity is such that it may be OBE at that point! I appreciate the feedback, and you taking a look.

Reply:
yeah that's huge. The worse thing is the unbounded size (if you happen to have more logs than you expect you can easily overflow available ram). Group by is actually expensive and should be used sparingly. I feel that in this case it can be done smarter by just deduplicating one file at a time since VSS is only looking at the same file across multiple vsc. For the backup use case we can probably deduplicate a glob of the backups (e.g. for each file look at the file plus the .bak extension - are all the backup files named consistently?)

Perhaps VQL needs a way to serialize to tempfiles when in memory queries are too large?

Reply:
Given the rarity of de-duplication outside VSS, a quick win would be to "factor it out" in the happy path, something like:

```sql
if(condition=SearchVSS, then={
  -- select with smarter, per-file GROUP BY
}, else={
  -- select without group by (the default)
}
```

along with a big warning about the possible VSS memory implications.

Reply:
Just to follow up: a quick test shows implementing it like this takes half the time and no memory spikes 😃

```sql

SELECT * FROM if(condition=SearchVSS, then={
            SELECT * FROM foreach(
              row=files,
              query={
                SELECT *,
                  timestamp(epoch=int(int=System.TimeCreated.SystemTime)) AS TimeCreated,
                  System.Channel as Channel,
                  System.EventRecordID as EventRecordID,
                  System.EventID.Value as EventID,
                  FullPath
                FROM parse_evtx(filename=FullPath)
                WHERE
                  if(condition=StartDate, then=TimeCreated >= timestamp(string=StartDate), else=true)
                  AND if(condition=EndDate, then=TimeCreated <= timestamp(string=EndDate), else=true)
                  AND Channel =~ ChannelRegex
                  AND str(str=EventID) =~ IDRegex
              }
            )
            GROUP BY EventRecordID,Channel
         }, else={
            SELECT * FROM foreach(
              row=files,
              query={
                SELECT *,
                  timestamp(epoch=int(int=System.TimeCreated.SystemTime)) AS TimeCreated,
                  System.Channel as Channel,
                  System.EventRecordID as EventRecordID,
                  System.EventID.Value as EventID,
                  FullPath
                FROM parse_evtx(filename=FullPath)
                WHERE
                  if(condition=StartDate, then=TimeCreated >= timestamp(string=StartDate), else=true)
                  AND if(condition=EndDate, then=TimeCreated <= timestamp(string=EndDate), else=true)
                  AND Channel =~ ChannelRegex
                  AND str(str=EventID) =~ IDRegex
              }
            ) 
         })
```

Reply:
Cool! Maybe refactor the query into a stored query so it is more compact?

```sql
LET EVTXSearch =  SELECT * FROM foreach(
              row=files,
              query={
                SELECT *,
                  timestamp(epoch=int(int=System.TimeCreated.SystemTime)) AS TimeCreated,
                  System.Channel as Channel,
                  System.EventRecordID as EventRecordID,
                  System.EventID.Value as EventID,
                  FullPath
                FROM parse_evtx(filename=FullPath)
                WHERE
                  if(condition=StartDate, then=TimeCreated >= timestamp(string=StartDate), else=true)
                  AND if(condition=EndDate, then=TimeCreated <= timestamp(string=EndDate), else=true)
                  AND Channel =~ ChannelRegex
                  AND str(str=EventID) =~ IDRegex
              }
            )

SELECT * FROM if(condition=SearchVSS, then={
    SELECT * FROM EVTXSearch
     GROUP BY EventRecordID,Channel
}, else=EVTXSearch)
```

Reply:
@scudette this is true across all EVTX with VSS and other logs. My recent MEGASync logs would have the same issue with VSS too as that can be a large rotated log set. I cut down the regex filter but you could set it to "." and cause a lot of rows.

Maybe we need to allow the VSS hunt to dedup on much smaller result sets with a set LIMIT or just not dedup?

Matt